### DISCUSSION TITLE
parseStringLiteral treats single quote and double quoted strings the same wrt escaping

### DISCUSSION THREAD

**Discussion by nhoughto** [harttle/liquidjs#289] [Q&A]
I'm working through a difference in input handling between this and other liquid libraries, specifically wrt string literals, in this example:

```
{{ value | replace :  '"' '\"' }}
```

These two string literals are interpreted by JS as 
1. `["'", "\"", "'"]` aka single quote, escaped double quote, single quote in a 3 char string
2. `["'", "\\", "\"", "'"]` aka single quote, escaped backslash, escaped quote, single quote in a 4 char string

the two pieces are parsed by `parseStringLiteral` excerpt, my read of this is that we skip the first char being the quote (!) and then treat proceeding escape chars the same whether the string literal uses single or double quotes, to do escaping as expected there should be two paths a single quote path and a double quote path? If started with single quote, then `\"` is two chars, if started with double quote then `\"` is one char (escaped quote).

```
...
for (let i = 1; i < str.length - 1; i++) {
    if (str[i] !== '\\') {
      ret += str[i]
      continue
    }
...
```

This has the impact in my simple example of making the replace filter never fire, as the parseStringLiteral method considers these two strings equivalent.

**Comment by harttle** [comment]
Observations for shopify/liquid, and decisions in LiquidJS:

- in shopify/liquid, both `'` and `"` are terminated by the next `'` and `"` respectively, even if there's a `\` before the ending quote. LiquidJS won't inherit this bug.
- in shopify/liquid, `'` and `"` are treated the same, not in the Ruby way. LiquidJS will do the same since JavaScript also happens to treat both single and double quote the same.

To clarify, the following test snippets both render to `"b"` in LiquidJS whereas the last one renders differently in shopify/liquid, which is `b` .

```liquid
{{ "aba" | replace :  'a', '"' }}
{{ "aba" | replace :  'a', '\"' }}
{{ "aba" | replace :  'a', "\"" }}
```
